lib/AST/Type.cpp

```diff
@@ -2474,12 +2474,36 @@ getArchetypeAndRootOpaqueArchetype(Type maybeOpaqueType) {
 
 bool ReplaceOpaqueTypesWithUnderlyingTypes::shouldPerformSubstitution(
     OpaqueTypeDecl *opaque) const {
-  return shouldPerformSubstitution(opaque, context);
+  return shouldPerformSubstitution(opaque, contextModule, contextExpansion);
 }
 
-static Type substOpaqueTypesWithUnderlyingTypes(
-    Type ty, SILFunction *context) {
-  ReplaceOpaqueTypesWithUnderlyingTypes replacer(context);
+bool ReplaceOpaqueTypesWithUnderlyingTypes::shouldPerformSubstitution(
+    OpaqueTypeDecl *opaque, ModuleDecl *contextModule,
+    ResilienceExpansion contextExpansion) {
+  auto namingDecl = opaque->getNamingDecl();
+
+  // Allow replacement of opaque result types of inlineable function regardless
+  // of resilience and in which context.
+  if (namingDecl->getAttrs().hasAttribute<InlinableAttr>()) {
```

Comment:
This should be namingDecl->getResilienceExpansion() == ResilienceExpansion::Minimal to correctly handle functions nested inside inlinable functions, accessors of inlinable properties ,etc

Reply:
I am confused.

I thought I have to look at the resilience expansion of the context (the function in which I perform the substitution). This happens on the next lines.

The namingDecl is the decl that has the opaque result type. If the namingDecl function is inlinable and from a different resilient module it should still be okay to replace the result type.

```
ResilientModule.swift

struct ResilientStruct {
     @inlineable
     func namingDecl() -> some P {
         return Int64(1)
    }
}

ContextModule.swift


func context() {
    let r = ResilientStruct()
    let opaque = r.namingDecl() // Okay to specialize to Int64 here
}
```

Reply:
Here are the rules I have currently implemented (or believe to have implemented ;-) :

- if the function with the opaque result type is inline-able allow substitution
(see example above)
- if the context function in which we perform substitution has maximal resilience expansion and the context function and the opaque type's function are in the same module allow substitution
- if the opaque types's module is not resilient allow substitution
- otherwise disallow substitution

Reply:
Sorry about the confusion. The right way to check if namingDecl is inlineable is "namingDecl->getResilienceExpansion() == ResilienceExpansion::Minimal ". You should never check for the presence of InlineableAttr directly.

Reply:
Ah. Okay. Now I got it. Makes sense.

https://github.com/apple/swift/pull/24478